src/Chapter3.hs

```diff
@@ -1007,6 +1135,145 @@ contestants, and write a function that decides the outcome of a fight!
 -}
 
 
+
+{-
+My doctests:
+
+>>> :{
+  fight'
+    FightKnight (FightHealth 10) (FightAttack 2) (FightDefense 0) [KnightAttack, KnightDrinkPotion 1]
+    FightMonster (FightHealth 10) (FightAttack 3) [MonsterAttack]
+:}
+Victory
+
+>>> :{
+  fight'
+    FightKnight (FightHealth 3) (FightAttack 2) (FightDefense 0) [KnightAttack, KnightCastSpell 1]
+    FightMonster (FightHealth 10) (FightAttack 3) [MonsterAttack]
+:}
+Defeat
+
+>>> :{
+  fight'
+    FightKnight (FightHealth 3) (FightAttack 2) (FightDefense 0) [KnightAttack]
+    FightMonster (FightHealth 10) (FightAttack 3) [MonsterRun]
+:}
+Victory
+
+-}
+
+newtype FightHealth = FightHealth Int deriving (Show, Eq)
+newtype FightAttack = FightAttack Int deriving (Show, Eq)
+newtype FightDefense = FightDefense Int deriving (Show, Eq)
+
+data FightAction
+  = KnightAttack
+  | KnightDrinkPotion Int
+  | KnightCastSpell Int
+  | MonsterAttack
+  | MonsterRun
```

Comment:
Nice sum type!

But I think it could be more type-safe, if we would have Monster and Knight actions separately. In that case, we will restrict the possible state to only relevant ones. 🙂